### ISSUE TITLE
Check whether hledger-web is running within Sandstorm

### ISSUE THREAD

**Issue by ndarilek** [simonmichael/hledger#327]
For the Sandstorm port, it would be useful to do the following:
- Check whether hledger-web is running within a Sandstorm environment. Traditionally, Sandstorm packages set a `SANDSTORM` environment variable. You might check for that, or add a `--sandstorm` switch that runs hledger-web in Sandstorm mode regardless of the environment.
- Check Sandstorm permissions. These are passed in an `X-Sandstorm-Permissions` HTTP header, and should only be checked if Sandstorm support is enabled. I think they're comma-delimited but the docs don't say for sure. I'm sending a request for clarification and will update once I know more.

Once these two features have landed, we can potentially build other features on top. Some I'd like to see:
- Separate write permission. Should just be a matter of checking the relevant permission, not displaying relevant UI functionality when it isn't present, and checking it on any attempts to modify the journal.
- Some sort of mass read/write functionality for downloading/uploading entire journals. Good for seeding a Sandstorm grain (app instance) with existing journal content.
- Ability to import/export CSV, integrated with permissions. I think I read in another issue that this might be available in code but hasn't been integrated into hledger-web yet. Generally we'll need easy import/export because you don't have easy command line access to a grain's content like you do with hledger-web running locally.

Thanks. I can probably help with some of these if we have a nice way to check for Sandstorm/permissions, but I have a hard time wrapping my brain around Haskell. :)
##

**Comment by simonmichael** [comment]
This sounds good.

We used to have a whole-journal edit form which I disabled, it may still be in the code.

[CSV export](http://hledger.org/manual.html#csv-output) should be easy to expose in the web UI. [CSV import](http://hledger.org/manual.html#csv) is a bigger project, it will probably some sort of UI for configuring the CSV conversion rules (or at least for uploading rules you've tested locally).

For permissions, which we should have with or without sandstorm, how about these levels:

- view - view all data, export
- add - also add new transactions (appends to the journal)
- edit - also change or delete existing transactions, one at a time
- manage - also edit the whole file, import, change settings etc.

**Comment by ndarilek** [comment]
Looks good. Regarding permissions, I'd recommend getting rid of `view`. 
With Sandstorm, no one can view anything unless you've shared it with 
them, so that case is simply handled by having no permissions.

Thanks!

**Comment by simonmichael** [comment]
Well call them "roles" then - the app should have some notion of what role the current user is.

**Comment by ndarilek** [comment]
It does. The default "viewer" role would just have no permissions. 
Hledger-web would just check for "write" and any others we'd define, 
which would be managed by Sandstorm-defined roles.

**Comment by ndarilek** [comment]
Confirmed, Sandstorm permissions are comma-separated in the header.

**Comment by paulproteus** [comment]
Hi @simonmichael & @ndarilek ,

I ran into Bradley Kuhn, an avid ledger user, last week at Debconf, which reminded me of hledger on Sandstorm. For context, I'm one of the developers of Sandstorm, and the main maintainer of the Sandstorm docs.

I noticed that @simonmichael responded to a comment on the HLedger Web page on the Sandstorm App Market, and I was impressed at how earnest and thoughtful and helpful the response was! So I was wondering what makes sense for the future of this package, and if maybe it makes sense for you to be the person who pushes updates to the Sandstorm app market. Nolan is also kind and graceful, but I figure two maintainers might be better than one. If so, perhaps Nolan would be willing to share the app key with you. I'll leave that to you two to decide, but I wanted to suggest it.

As for this specific issue  - any chance you're interested in looking into it in the near future? Anything I can do to help or express my gratitude?

Thanks!

**Comment by simonmichael** [comment]
Thanks @paulproteus! Sandstorm is terrific. I'm not using hledger's web interface much myself at the moment and many other things need attention so this is not reaching the top of my queue. I've marked it as a requirement for 1.0 if not sooner. I wonder how to raise the profile of hledger-web on Sandstorm and generally, to attract more help.

**Comment by simonmichael** [comment]
@ndarilek: https://dev.thewordnerd.info/nolan/hledger seems offline, has it moved ?

**Comment by ocdtrekkie** [comment]
If HLedger Web currently doesn't support roles in itself, there's not much point in Sandstorm permissions being detected at this juncture.

Import/export is a big topic, and probably the top one to resolve for #425, though if we are mostly concerned with making the experience good for HLedger users, do we "just" need to import/export the Ledger file as-is? An HLedger user might export it to make adjustments not possible via the web interface, and then re-import it to update their web/cloud-stored version.

**Comment by zarybnicky** [comment]
@simonmichael I'm starting to work on this, and the first thing I see there is a ton of CPP (e.g. http-conduit or yesod versions). How is it with backwards compatibility? I see that Travis attempts to build using GHC 8.4.3 and Appveyor using GHC 8.2 (and is currently failing due to missing `happy`; also, the links `travis.hledger.org` etc. from README are currently not working via https).

How much of this can I remove? Should hledger-web be buildable using `stack-ghc710.yaml` all the way through `stack.yaml`? If so, I can safely remove at least the CPP related to http-conduit, as its version is >2 in LTS 6.35 as well. Not yesod though, it's on version 1.4 there, which is before the big HandlerT refactoring.

**Comment by simonmichael** [comment]
@zarybnicky, great. We should build warning free with all the stack*.yaml, though ghc7.10 is a stretch goal and can be dropped if it gives a lot of trouble. make buildplan-test-all and make build-test-all will test this. I will do this in the travis script soon if disk and time quota allow. Yes do remove any CPP we can do without.

I've seen those missing happy errors which I think are recent; it's worth trying latest stack (stack upgrade --git) or older ghc (8.4.2). 

IIRC way back then I tried to adjust yesod's file layout to be easier to work with. Feel free to try the standard layout and we can compare.

**Comment by zarybnicky** [comment]
@simonmichael: Having worked in this for a while, I've made quite a few changes (I'll try to open a WIP pull-request in a bit), and fixed some oversights - starting with #433.

I'm trying to solve a few other along the way, so I have a few questions:

* regarding editing the journal - the 'edit' form I found is just a big textarea of the entire journal. Ideally I'd like to have an edit field per transaction - I haven't really looked through hledger-lib yet, but is there support for this?
* regarding interactivity (e.g. infinite scrolling, or AJAX forms) - I'd like to use Yesod's TypedContent functionality and communicate also via JSON. Should we care about hledger-web's usability on JS-disabled browsers? Currently, browsing works so-so, but adding transactions doesn't, so I'd like to add a bit more JS-only functionality.
* also regarding JSON - is there any particular reason hledger uses Galois' `json` instead of the now standard `aeson`? In order to use the TypedContent mechanism, I think I'll have to add aeson as a dependency anyway (not sure, I haven't used it in a while).
